### ISSUE TITLE
Use spack to facilitate bug reporting

### ISSUE THREAD

**Issue by markcmiller86** [spack/spack#5248]
I culled this from my orig. comments in #2532

My first thought for this is for failures installing and/or testing a package.

There are two pieces to this suggestion...

1. Assembling package install/use provenance for a fully detailed bug report
    - package version information
    - patches applied
    - variants settings
    - dependencies and their versions and variants
    - OS and compiler info, etc.
    - Failures during tests/checks could also include test stdout/stderr
1. Getting the resulting information to the right people
    - the right people are the developers of the package Spack is installing
        - These may or may not be the same people who wrote Spack's `package.py` (e.g. the _maintainers_) file (see #5230)
    - probably via some email
    - package.py file would need something like `report-bugs(<some-email-address>` field

Step one could be done automatically maybe as part of a `spack install --assemble-bug-report` command or something. In other words, it could be something that is requested as part of an install. Step two could be handled either manually or, if package's know their bug-report email addresses, by an additional spack command or `spack install` subcommand.

**Comment by mathstuf** [comment]
It'd be great if it just dumped out the required information in markdown, ready for pasting into a new issue (with a link). Logs should probably go into their own files and uploaded, but I think that is going to be a manual step. There is now #2532 which can automatically add `@mention` bits to the relevant users as well.

**Comment by davydden** [comment]
> Assembling package install/use provenance for a fully detailed bug report

i think homebrew has a handy command `brew gist-logs <formula>` which gathers all the needed info and uploads it as a gist.

> Getting the resulting information to the right people

I think there are two options here: 

1. The issue is with the package itself and not with how it is being installed via Spack (i.e. bug with AVX SIMD when using compiler Y). Then the right people are package developers. To that we probably need another _section_  to report package specific bugs to. I would say it should be either a URL or email. URL's could point to mailing lists or GitHub issues (i.e. https://www.mcs.anl.gov/petsc/miscellaneous/mailing-lists.html or https://github.com/trilinos/Trilinos/issues ).

I don't think Spack should/can handle this step automatically in any way. But it is good to have a quick link/email ready for Spack users so that they know where to report.

2. When the problem is with installing the package via Spack. Then _maintainers_ https://github.com/LLNL/spack/pull/5230 are the right people to ping. If there are no maintainers attached, one can look for package authors.

Unfortunately, it's not always clear whether a problem is from category (1) or (2).

p.s. Generally, package maintainers != package authors != package developers.

**Comment by mathstuf** [comment]
Yes, but the way distributors (like Fedora, Debian, etc.) do this normally is that the maintainer gets the issue and either forwards it themselves to upstream or gets the user going to the right place for communicating with upstream.

**Comment by davydden** [comment]
> the maintainer gets the issue and either forwards it themselves to upstream or gets the user going to the right place for communicating with upstream.

oh, i did not know that. Makes sense as well. Then theoretically we should be fine only with `maintainers` field.

**Comment by markcmiller86** [comment]
FWIW...I don't like the idea that a consequence of adding a package to Spack (thereby becoming a _maintainer_ of that package), is that I get the additional burden of being a _go between_ for marshalling bug reports to package developers. That isn't a scalable solution. And, if all it is going to entail is that a _maintainter_ turns around and forwards the bug-report to package developers, I see no reason why the email address can't  just be put into the package.py file allowing users to discover and use it themselves.

**Comment by markcmiller86** [comment]
So, I guess the real issue is that when problems occur, its impossible for users to know if the problem is one a _maintainer_ would fix or one a _developer_ would fix.

**Comment by davydden** [comment]
>  consequence of adding a package to Spack (thereby becoming a maintainer of that package)

A bit tangential from my side, but adding a package and becoming a maintainer are two different things. You become maintainer voluntarily by adding yourself in `maintainers` field. You don't have to, even for those packages you added. I added a dozens of packages to Spack, but I only added myself as a maintainer of `dealii`.

But I see your point and for upstream bugreports one could still have some URL or email in the package itself. I guess Spack as a community needs to decide between those two models.

@tgamblin @alalazo @adamjstewart ping.

> when problems occur, its impossible for users to know if the problem is one a maintainer would fix or one a developer would fix.

yes, i fully agree with that.
